Hello @tchap,

I have added a simple unit test to prove the mitigation and show how this problem can happen:

### 1. Test Run _Without_ Mitigation

```
=== RUN   TestTrie_compactChildrenHeadIsNil
=== RUN   TestTrie_compactChildrenHeadIsNil/children_head_is_nil
--- FAIL: TestTrie_compactChildrenHeadIsNil/children_head_is_nil (0.00s)

--- FAIL: TestTrie_compactChildrenHeadIsNil (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x18 pc=0x1049d0e58]

goroutine 19 [running]:
testing.tRunner.func1.2({0x104a2f000, 0x104b2fbd0})
	/Users/atakan.pehlivanoglu/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.10.darwin-arm64/src/testing/testing.go:1632 +0x1bc
testing.tRunner.func1()
	/Users/atakan.pehlivanoglu/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.10.darwin-arm64/src/testing/testing.go:1635 +0x334
panic({0x104a2f000?, 0x104b2fbd0?})
	/Users/atakan.pehlivanoglu/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.10.darwin-arm64/src/runtime/panic.go:791 +0x124
github.com/tchap/go-patricia/v2/patricia.(*Trie).compact(0x14000132140)
	/Users/atakan.pehlivanoglu/Documents/Bitbucket/go-patricia/patricia/patricia.go:468 +0x58
github.com/tchap/go-patricia/v2/patricia.TestTrie_compactChildrenHeadIsNil.func1(0x1400011a9c0)
	/Users/atakan.pehlivanoglu/Documents/Bitbucket/go-patricia/patricia/patricia_sparse_test.go:642 +0x50
testing.tRunner(0x1400011a9c0, 0x104a51990)
	/Users/atakan.pehlivanoglu/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.10.darwin-arm64/src/testing/testing.go:1690 +0xe4
created by testing.(*T).Run in goroutine 18
	/Users/atakan.pehlivanoglu/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.10.darwin-arm64/src/testing/testing.go:1743 +0x314

Process finished with the exit code 1
```

### 2. Test Run _With_ Mitigation
```

=== RUN   TestTrie_compactChildrenHeadIsNil
=== RUN   TestTrie_compactChildrenHeadIsNil/children_head_is_nil
--- PASS: TestTrie_compactChildrenHeadIsNil/children_head_is_nil (0.00s)
--- PASS: TestTrie_compactChildrenHeadIsNil (0.00s)
PASS

Process finished with the exit code 0
```

--- a/patricia/patricia_sparse_test.go
+++ b/patricia/patricia_sparse_test.go
@@ -9,6 +9,7 @@ import (
 	"bytes"
 	"errors"
 	"fmt"
+	"reflect"
 	"strings"
 	"testing"
 )
@@ -628,6 +629,19 @@ func TestTrie_compact(t *testing.T) {
 			}
 		}
 		return nil
+	})
+}
+
+func TestTrie_compactChildrenHeadIsNil(t *testing.T) {
+	testName := "children head is nil"
+	t.Run(testName, func(t *testing.T) {
+		trie := NewTrie()
+		trie.children.add(nil)
+
+		// should return trie itself
+		if got := trie.compact(); !reflect.DeepEqual(got, trie) {
+			t.Errorf("compact() = %v, want %v", got, trie)
+		}
 	})
 }
 

